src/lib.rs

```diff
@@ -27,33 +38,66 @@ pub extern "C" fn wapc_init() {
     register_function("protocol_version", protocol_version_guest);
 }
 
+// Represents an abstraction of an struct that contains an image
+// Used to reuse code for Container and EphemeralContainer
+trait ImageHolder: Clone {
+    fn set_image(&mut self, image: Option<String>);
+    fn get_image(&self) -> Option<String>;
+}
+
+impl ImageHolder for Container {
+    fn set_image(&mut self, image: Option<String>) {
+        self.image = image;
+    }
+
+    fn get_image(&self) -> Option<String> {
+        self.image.clone()
+    }
+}
+
+impl ImageHolder for EphemeralContainer {
+    fn set_image(&mut self, image: Option<String>) {
+        self.image = image;
+    }
+
+    fn get_image(&self) -> Option<String> {
+        self.image.clone()
+    }
+}
+
 fn validate(payload: &[u8]) -> CallResult {
     let validation_request: ValidationRequest<Settings> = ValidationRequest::new(payload)?;
-
     info!(LOG_DRAIN, "starting validation");
 
-    // TODO: you can unmarshal any Kubernetes API type you are interested in
-    match serde_json::from_value::<apicore::Pod>(validation_request.request.object) {
-        Ok(pod) => {
-            // TODO: your logic goes here
-            if pod.metadata.name == Some("invalid-pod-name".to_string()) {
-                let pod_name = pod.metadata.name.unwrap();
-                info!(
-                    LOG_DRAIN,
-                    "rejecting pod";
-                    "pod_name" => &pod_name
-                );
-                kubewarden::reject_request(
-                    Some(format!("pod name {} is not accepted", &pod_name)),
-                    None,
-                )
-            } else {
-                info!(LOG_DRAIN, "accepting resource");
-                kubewarden::accept_request()
+    match serde_json::from_value::<apicore::Pod>(validation_request.request.object.clone()) {
+        Ok(mut pod) => {
+            if let Some(spec) = pod.spec {
+                match verify_all_images_in_pod(&spec, &validation_request.settings.signatures) {
+                    Ok(spec_with_digest) => {
+                        if validation_request.settings.modify_images_with_digest {
+                            pod.spec = Some(spec_with_digest);
+                            let mutated_object = serde_json::to_value(&pod)?;
+                            return kubewarden::mutate_request(mutated_object);
```

Comment:
This assumes the users never reference images by digest. This could lead to mutation requests responses that are not changing the Pod. This is not causing harm, but I think it would be better to reply with a `mutate_request` only when the contents of the Pods have been actually changed

Reply:
yes, that makes sense. done in https://github.com/kubewarden/verify-image-signatures/pull/4/commits/6cf9839979d4e5e78642006d40e7c373a65f4418